Fails to build with Boost 1.69 (develop snapshot)

After boostorg/logic@23cd89d4c80f build fails. See [error log](https://ptpb.pw/aK-J).

```c++
src/CSGTreeEvaluator.cc:179:9: error: no viable conversion from 'boost::tribool' to 'bool'
                        bool convex = ps->convexValue();
                             ^        ~~~~~~~~~~~~~~~~~
```


<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65048990-fails-to-build-with-boost-1-69-develop-snapshot?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github).
</bountysource-plugin>

> Sounds good! Would you mind opening a Pull Request for that fix?

That maybe hard. `master` requires Qt5QScintilla which (unlike Qt4 one) isn't packaged on FreeBSD. Commenting out 
https://github.com/openscad/openscad/blob/8912e648df2f84b42d7406f5d2a50218f72f8ab6/CMakeLists.txt#L131-L134
leads to an unrelated bustage ([fixed downstream](https://github.com/freebsd/freebsd-ports/blob/1548b5448a3fa91810be061182d28c3cd070d772/cad/openscad-devel/files/patch-src-ModuleCache.h))
```c++
In file included from src/ModuleCache.cc:1:
src/ModuleCache.h:14:2: error: unknown type name 'time_t'
        time_t evaluate(const std::string &filename, class FileModule *&module);
        ^
src/ModuleCache.h:30:3: error: unknown type name 'time_t'
                time_t mtime{};          // time file last modified
                ^
src/ModuleCache.h:31:3: error: unknown type name 'time_t'
                time_t includes_mtime{}; // time the includes last changed
                ^
In file included from src/ModuleCache.cc:1:
In file included from src/ModuleCache.h:4:
In file included from /usr/include/c++/v1/unordered_map:369:
/usr/include/c++/v1/__hash_table:2038:29: error: cannot initialize object parameter of type 'std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *> *>' with an expression of type 'std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *>'
            __pn->__next_ = __h.get()->__ptr();
                            ^~~~~~~~~
/usr/include/c++/v1/unordered_map:1386:21: note: in instantiation of function template specialization 'std::__1::__hash_table<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::__unordered_map_hasher<std::__1::basic_string<char>, std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::hash<std::__1::basic_string<char> >, true>, std::__1::__unordered_map_equal<std::__1::basic_string<char>, std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::equal_to<std::__1::basic_string<char> >, true>, std::__1::allocator<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry> > >::__emplace_unique_key_args<std::__1::basic_string<char>, const std::__1::piecewise_construct_t &, std::__1::tuple<const std::__1::basic_string<char> &>, std::__1::tuple<> >' requested here
    return __table_.__emplace_unique_key_args(__k,
                    ^
src/ModuleCache.cc:53:41: note: in instantiation of member function 'std::__1::unordered_map<std::__1::basic_string<char>, ModuleCache::cache_entry, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, ModuleCache::cache_entry> > >::operator[]' requested here
        cache_entry &cacheEntry = this->entries[filename];
                                               ^
In file included from src/ModuleCache.cc:1:
In file included from src/ModuleCache.h:4:
In file included from /usr/include/c++/v1/unordered_map:369:
/usr/include/c++/v1/__hash_table:2043:23: error: cannot initialize object parameter of type 'std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *> *>' with an expression of type 'std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *>'
                    = __h.get()->__ptr();
                      ^~~~~~~~~
/usr/include/c++/v1/__hash_table:2050:16: error: static_cast from 'std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *>, std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *> > > >::pointer' (aka 'std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *> *') to 'std::__1::__hash_table<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::__unordered_map_hasher<std::__1::basic_string<char>, std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::hash<std::__1::basic_string<char> >, true>, std::__1::__unordered_map_equal<std::__1::basic_string<char>, std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, std::__1::equal_to<std::__1::basic_string<char> >, true>, std::__1::allocator<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry> > >::__next_pointer' (aka 'std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char>, ModuleCache::cache_entry>, void *> *> *'), which are not related by inheritance, is not allowed
        __nd = static_cast<__next_pointer>(__h.release());
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
6 errors generated.
```

--- a/openscad.pro
+++ b/openscad.pro
@@ -575,7 +575,7 @@ unix:!macx {
   SOURCES += src/input/DBusInputDriver.cc
 }
 
-unix:!macx {
+linux: {
   DEFINES += ENABLE_JOYSTICK
 
   HEADERS += src/input/JoystickInputDriver.h

--- a/openscad.pro
+++ b/openscad.pro
@@ -565,7 +565,7 @@ HEADERS += src/ext/libtess2/Include/tesselator.h \
            src/ext/libtess2/Source/sweep.h \
            src/ext/libtess2/Source/tess.h
 
-unix:!macx {
+has_qt5:unix:!macx {
   QT += dbus
   DEFINES += ENABLE_DBUS
   DBUS_ADAPTORS += org.openscad.OpenSCAD.xml

--- a/src/CSGTreeEvaluator.cc
+++ b/src/CSGTreeEvaluator.cc
@@ -176,7 +176,7 @@ shared_ptr<CSGNode> CSGTreeEvaluator::evaluateCSGNodeFromGeometry(
 			auto ps = dynamic_pointer_cast<const PolySet>(geom);
 			// Since is_convex() doesn't handle non-planar faces, we need to tessellate
 			// also in the indeterminate state so we cannot just use a boolean comparison. See #1061
-			bool convex = ps->convexValue();
+			bool convex{ps->convexValue()};
 			if (ps && !convex) {
 				assert(ps->getDimension() == 3);
 				auto ps_tri = new PolySet(3, ps->convexValue());

--- a/src/FileModule.h
+++ b/src/FileModule.h
@@ -3,7 +3,7 @@
 #include <string>
 #include <unordered_map>
 #include <unordered_set>
-#include <time.h>
+#include <ctime>
 
 #include "module.h"
 #include "value.h"
@@ -23,8 +23,8 @@ public:
 	const std::string &modulePath() const { return this->path; }
 	void registerUse(const std::string path);
 	void registerInclude(const std::string &localpath, const std::string &fullpath);
-	time_t includesChanged() const;
-	time_t handleDependencies();
+	std::time_t includesChanged() const;
+	std::time_t handleDependencies();
 	bool hasIncludes() const { return !this->includes.empty(); }
 	bool usesLibraries() const { return !this->usedlibs.empty(); }
 	bool isHandlingDependencies() const { return this->is_handling_dependencies; }
@@ -40,7 +40,7 @@ private:
 		std::string filename;
 	};
 
-	time_t include_modified(const IncludeFile &inc) const;
+	std::time_t include_modified(const IncludeFile &inc) const;
 
 	typedef std::unordered_map<std::string, struct IncludeFile> IncludeContainer;
 	IncludeContainer includes;

--- a/src/FontCache.h
+++ b/src/FontCache.h
@@ -29,7 +29,7 @@
 #include <string>
 #include <iostream>
 
-#include <time.h>
+#include <ctime>
 
 #include <ft2build.h>
 #include FT_FREETYPE_H
@@ -93,7 +93,7 @@ public:
     static void registerProgressHandler(InitHandlerFunc *handler, void *userdata = nullptr);
 
 private:
-    typedef std::pair<FT_Face, time_t> cache_entry_t;
+    typedef std::pair<FT_Face, std::time_t> cache_entry_t;
     typedef std::map<std::string, cache_entry_t> cache_t;
 
     static FontCache *self;

--- a/src/ModuleCache.cc
+++ b/src/ModuleCache.cc
@@ -9,7 +9,6 @@
 #include <stdio.h>
 #include <fstream>
 #include <sstream>
-#include <time.h>
 #include <sys/stat.h>
 #include <algorithm>
 
@@ -29,7 +28,7 @@ ModuleCache *ModuleCache::inst = nullptr;
 
 	Returns the latest modification time of the module, its dependencies or includes.
 */
-time_t ModuleCache::evaluate(const std::string &filename, FileModule *&module)
+std::time_t ModuleCache::evaluate(const std::string &filename, FileModule *&module)
 {
 	module = nullptr;
 	auto entry = this->entries.find(filename);
@@ -67,7 +66,7 @@ time_t ModuleCache::evaluate(const std::string &filename, FileModule *&module)
 			shouldCompile = false;
 			// Recompile if includes changed
 			if (cacheEntry.parsed_module) {
-				time_t mtime = cacheEntry.parsed_module->includesChanged();
+				std::time_t mtime = cacheEntry.parsed_module->includesChanged();
 				if (mtime > cacheEntry.includes_mtime) {
 					cacheEntry.includes_mtime = mtime;
 					shouldCompile = true;
@@ -116,7 +115,7 @@ time_t ModuleCache::evaluate(const std::string &filename, FileModule *&module)
 	
 	module = lib_mod;
 	// FIXME: Do we need to handle include-only cases?
-	time_t deps_mtime = lib_mod ? lib_mod->handleDependencies() : 0;
+	std::time_t deps_mtime = lib_mod ? lib_mod->handleDependencies() : 0;
 
 	return std::max({deps_mtime, cacheEntry.mtime, cacheEntry.includes_mtime});
 }

--- a/src/ModuleCache.h
+++ b/src/ModuleCache.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <string>
+#include <ctime>
 #include <unordered_map>
 
 /*!
@@ -11,7 +12,7 @@ class ModuleCache
 public:
 	static ModuleCache *instance() { if (!inst) inst = new ModuleCache; return inst; }
 
-	time_t evaluate(const std::string &filename, class FileModule *&module);
+	std::time_t evaluate(const std::string &filename, class FileModule *&module);
 	class FileModule *lookup(const std::string &filename);
 	bool isCached(const std::string &filename);
 	size_t size() { return this->entries.size(); }
@@ -27,8 +28,8 @@ private:
 		class FileModule *module{};
 		class FileModule *parsed_module{}; // the last version parsed for the include list
 		std::string cache_id;
-		time_t mtime{};          // time file last modified
-		time_t includes_mtime{}; // time the includes last changed
+		std::time_t mtime{};          // time file last modified
+		std::time_t includes_mtime{}; // time the includes last changed
 	};
 	std::unordered_map<std::string, cache_entry> entries;
 };

--- a/src/StatCache.cc
+++ b/src/StatCache.cc
@@ -27,9 +27,9 @@
 #include "StatCache.h"
 #include "printutils.h"
 
-#include <sys/timeb.h>
 #include <string>
 #include <unordered_map>
+#include <chrono>
 
 namespace {
 
@@ -37,9 +37,7 @@ const float stale = 0.190;  // Maximum lifetime of a cache entry chosen to be sh
 
 double millis_clock(void)
 {
-	struct timeb tb;
-	ftime(&tb);
-	return tb.time + double(tb.millitm) / 1000;
+	return std::chrono::duration_cast<std::chrono::duration<float>>(std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
 struct CacheEntry

--- a/src/parameter/ParameterWidget.cc
+++ b/src/parameter/ParameterWidget.cc
@@ -285,7 +285,7 @@ void ParameterWidget::onSetNameChanged(){
 		msgBox.setDefaultButton(QMessageBox::Cancel);
 
 		if (msgBox.exec() == QMessageBox::Cancel) {
-			comboBoxPreset->setCurrentText(oldName);
+			comboBoxPreset->lineEdit()->setText(oldName);
 		}else{
 			onSetDelete();
 		}

